src/ansible_navigator/actions/exec.py

```diff
@@ -0,0 +1,102 @@
+"""Run the :exec subcommand."""
+import os
+import shlex
+
+from typing import List
+from typing import Optional
+from typing import Tuple
+from typing import Union
+
+from . import _actions as actions
+from ..app import App
+from ..configuration_subsystem import ApplicationConfiguration
+from ..runner import Command
+
+
+@actions.register
+class Action(App):
+    """Run the :exec subcommand."""
+
+    # pylint: disable=too-few-public-methods
+
+    KEGEX = r"^e(?:xec)?$"
+
+    def __init__(self, args: ApplicationConfiguration):
+        """Initialize the action.
+
+        :param args: The current application configuration.
+        """
+        super().__init__(args=args, logger_name=__name__, name="exec")
+
+    @staticmethod
+    def _generate_command(exec_command: str, exec_shell: bool) -> Tuple[str, Optional[List[str]]]:
+        """Generate the command and args.
+
+        :param exec_command: The command to run
+        :param exec_shell: Should the command be wrapped in a shell
+        :returns: The command and any pass through arguments
+        """
+        pass_through_args = None
+        if exec_shell and exec_command:
+            command = "/bin/bash"
+            pass_through_args = ["-c", exec_command]
+        else:
+            parts = shlex.split(exec_command)
+            command = parts[0]
+            if len(parts) > 1:
+                pass_through_args = parts[1:]
+        return (command, pass_through_args)
+
+    def run_stdout(self) -> Union[None, int]:
+        """Run in mode stdout.
+
+        :returns: The return code or None
+        """
+        self._logger.debug("exec requested in stdout mode")
+        response = self._run_runner()
+        if response:
+            _, _, ret_code = response
+            return ret_code
+        return None
+
+    def _run_runner(self) -> Optional[Tuple]:
+        # pylint: disable=too-many-branches
+        # pylint: disable=too-many-statements
+        """Spin up runner.
+
+        :return: The stdout, stderr and return code from runner
+        """
+        if isinstance(self._args.set_environment_variable, dict):
+            set_envvars = {**self._args.set_environment_variable}
+        else:
+            set_envvars = {}
```

Comment:
Or did you mean this?
```suggestion
        set_envvars = (self._args.set_environment_variable or {}).copy() 
```

Reply:
Extra note on the var name — `set_envvars` implies an action, if I were to see it used somewhere in the code, I would assume that it needs to be called as `set_envvars()`. If it's supposed to hold data, it should use a noun and have an implication of being an object/a concept, not an action.

Reply:
the parameter in the settings is 'set_environment_variables` so that got carried around up until the point it's handed to runner. So oddly in this case, the name hooks back to the settings entry and wasn't intended to suggest a function name, but I agree with as a pattern.  I'll make a change here, but that won't eliminate it everywhere

Reply:
`self._args.set_environment_variable` can either be a Constant or a dict, I prefer doing a type check here instead, since it's not apparent that the check is for that, let me through something else in there and let me know what you think.

Reply:
@webknjaz it might be over the top, but let me know what you think

Reply:
@webknjaz any thoughts here? I'll leave this for you to resolve.

Reply:
Okay, keep an instance check if there's value in this information.

As for the var name, it's important to understand that the same piece of data may have different semantical meanings on different levels of abstraction — it's one thing on the calling side but another in a called function that operates lower-level primitives.